### DISCUSSION TITLE
Knockback Modifier

### DISCUSSION THREAD

**Discussion by tsuza** [nosoop/SM-TFCustAttr#5] [Created attributes]
# Knockback Modifier

This [plugin](https://github.com/Zabaniya001/tf2-knockback_modifier) lets you change the knockback of a weapon, as easy as that.

## How to apply the attribute

Shove it inside tf_custom_attributes.txt if you want to replace a normal weapon's model ( or anything else that supports custom attributes ) or in the Custom Attributes section inside a custom weapon's cfg if you use Custom Weapons X.

`"knockback modifier" 	"value"`

**value**: 1.0 is the default knockback. 0.5 would be half of the original knockback and 1.5 would be 50% more knockback. 
You can change it as much as you want, even remove it by putting 0.0.

Example: `0.4` ( 60% less knockback )

## Dependencies

[Dhooks](https://github.com/peace-maker/DHooks2)

[Custom Attributes Framework](https://github.com/nosoop/SM-TFCustAttr)

[TF2 Attributes](https://github.com/nosoop/tf2attributes)

You might need [this dependency as well](https://github.com/nosoop/stocksoup) if you want to compile it yourself.

#

If you find any issues let me know.

**Comment by tsuza** [comment]
# Quick window regarding how it is achieved

You can achieve this in three different, but still similar ways ( as in, they utilize CTFPlayer::ApplyPushFromDamage ).

**1 -** The insane method. Change the value of flDamageForceReduction by getting its offset ( I should have it somewhere ) and then... you might say load to that address, but it's an immediate, or whatever it is, so you'd have to write some code to make it jump to your own code. But as [Lord Scag](https://github.com/Scags/) has mentioned, you'd need lots of good scotch and a couple of nights, and some headache.

**2 -** The eTakeDamageInfo Info yoinker. Basically, you'd just need to store the damage value, scale it to your liking, load and call it a day.

**3 -** Last but not least, the "damage force increase hidden" attribute method. You just need to listen to pre and post of CTFPlayer::ApplyPushFromDamage. 
In the pre function, check if the weapon has the custom attribute. If it does get the current value of "damage force increase hidden", multiply it by the knockback modifier's value, set it as  TF2Attrib_AddCustomPlayerAttribute.
In the post function, check again if the weapon has the custom attribute. If it does get the value of "damage force increase hidden" and divide it by the knockback modifier's value ( to retrieve the original value ), remove the custom player attribute and then add it back.

At the end of the day, all of these methods work and achieve the same result, but the third method is the easiest.

**Reply by nosoop** [reply]
For fun, 3a would be to inject a new attribute with the `damage_force_reduction` attribute class, then it'd be an add / remove instead of a save / restore.  But that'd add a bit of complexity to the installation, and it wouldn't be any better for the attribute cache.

**Reply by tsuza** [reply]
Ohh, that actually sounds like a really cool idea. I didn't even think about that possibility. I might give it a try one day.
